### ISSUE TITLE
Issue with deploying/loading a custom resource making use of a CRD

### ISSUE THREAD

**Issue by Smith334** [kubernetes-client/java#2824]
**Describe the bug**
I am having an issue deploying a "Nginx VirtualServer" resource that makes use of a Custom Resource Definition (CRD). The CRD itself has no problems applying to the cluster, but deploying the custom VirtualServer resource just after throws an exception:

```
java.io.IOException: Unknown apiVersionKind k8s.nginx.org/v1/VirtualServer is it registered?
                at io.kubernetes.client.util.Yaml.modelMapper(Yaml.java:555)
                at io.kubernetes.client.util.Yaml.loadAll(Yaml.java:179)
                at io.kubernetes.client.util.Yaml.loadAll(Yaml.java:145)

```
The exception is coming from SnakeYaml itself not being able to map to an existing model.
I believe that there is a way to generate the Java CRD model beforehand, but is this actually necessary even for custom resources if I have deployed the CRD literally just before? Should the client not be aware of the definition at the point of trying to make use of the CRD? Unless it is and I am just messing something up.

This is the VirtualServer CRD: [https://github.com/nginxinc/kubernetes-ingress/blob/main/deployments/common/crds/k8s.nginx.org_virtualservers.yaml](url)

Every YAML resource, including the CRD, is deployed with this Java code:

```
  public void apply(String yaml) throws IOException, KubectlException {
    List<Object> all = Yaml.loadAll(yaml);
    apply(all);
  }

  public void apply(List<Object> all) throws KubectlException {
    for (Object object : all) {
      KubernetesObject source = (KubernetesObject) object;
      KubectlApply<KubernetesObject> apply = Kubectl.apply((Class<KubernetesObject>) source.getClass());
      apply.apiClient(ioClient());
      apply.resource(source);
      apply.name(source.getMetadata().getName());
      apply.execute();
    }
  }
```

This is the VirtualServer resource i am trying to deploy after deploying the VirtualServer CRD to the cluster:

```
# Ingress resources are bound to specific ingress controllers using an identifier
# Ingress Controller Identifier: nginx-firstone
kind: VirtualServer
apiVersion: k8s.nginx.org/v1
metadata: 
  name: nginx-firstone-dmz-vs
  namespace: dmz
spec: 
  host: some-host-name
  ingressClassName: nginx-firstone
  tls: 
    secret: firstone-secret
    redirect: 
      enable: true
      code: 301
  upstreams: 
  - name: psp-dmz-webinterfaceclient-service-80-us
    service: psp-dmz-webinterfaceclient-service
    port: 80
    connect-timeout: "60"
    send-timeout: "60"
    read-timeout: "60"
    keepalive: 75
  - name: psp-dmz-webinterfaceclient-service-85-us
    service: psp-dmz-webinterfaceclient-service
    port: 85
    connect-timeout: "60"
    send-timeout: "60"
    read-timeout: "60"
    keepalive: 75
  routes: 
  - path: /merchantportal
    action: 
      proxy: 
        upstream: psp-dmz-webinterfaceclient-service-80-us
        requestHeaders: 
            pass: true
        rewritePath: /merchantportal
    errorPages: 
      - codes: [404]
        return: 
          code: 200
          body: This was returned as a 200
  - path: /merchantenrolment
    action: 
      proxy: 
        upstream: psp-dmz-webinterfaceclient-service-85-us
        requestHeaders: 
            pass: true
        rewritePath: /merchantenrolment
    errorPages: 
      - codes: [404]
        return: 
          code: 200
          body: This was returned as a 200
---
```
A few months back, I was able to deploy the CRD and the VirtualServer resource with no problem, and no need for any model generations.

**Client Version**
e.g. `1.19`

**Kubernetes Version**
e.g. `1.27`

**Java Version**
e.g. Java 11

**To Reproduce**
Simply deploy the CRD first and then deploy a VirtualServer resource that makes use of it. An exception will throw.

**Expected behavior**
The following exception will be thrown when attempting to deploy the VirtualServer resource:
```
java.io.IOException: Unknown apiVersionKind k8s.nginx.org/v1/VirtualServer is it registered?
                at io.kubernetes.client.util.Yaml.modelMapper(Yaml.java:555)
                at io.kubernetes.client.util.Yaml.loadAll(Yaml.java:179)
                at io.kubernetes.client.util.Yaml.loadAll(Yaml.java:145)
```

**Server (please complete the following information):**
 - OS: Windows
 - Environment: Windows VM
 - Cloud: AWS

**Additional context**
A few months back, I was able to deploy the CRD and the VirtualServer resource with no problem, and no need for any model generations.

**Comment by brendandburns** [comment]
The YAML loaded needs to have a mapping from a Java class to a the api group/version/kind. For CRDs you need to establish that mapping yourself. You can use the `ModelMapper` class to do this.

Hope that helps!

**Comment by brendandburns** [comment]
There is also the `Discovery` class which tries to do this dynamically, you can call discovery after the CRD is added to the cluster, but you still need a Java class to map to the CRD.

**Comment by Smith334** [comment]
Thanks for the reply @brendandburns . Is there any desire to allow for these POJOs to be automatically generated so that this mapping can take place without having to do any extra steps when using CRDs?

Unless you know of a way to prevent SnakeYaml from parsing/mapping the Yaml, and instead just deploy it to the Cluster using this client? I know for a fact that the CRD has already been deployed to the Cluster, and is working.

**Comment by brendandburns** [comment]
The automatic generation of code is the CRD generation process as describe here:
https://github.com/kubernetes-client/java/wiki/5.-Generate-Java-CRD-Model

But this is static code generation, you would do it before you deploy your code.

For dynamic code generation, the right thing to do is to write a custom YAML (de)serializer for DynamicKubernetesObject, there's an example serializer here: 

https://github.com/kubernetes-client/java/blob/master/util/src/main/java/io/kubernetes/client/util/Yaml.java#L287

That you could potentially use as a model.

Basically you could map from the api group/kind -> DynamicKubernetesObject and parse the YAML into that.

If you want to just send bare YAML as utf-8 instead of parsing the YAML, you could look at the code here:

https://github.com/kubernetes-client/java/blob/master/kubernetes/src/main/java/io/kubernetes/client/openapi/ApiClient.java#L1009

That's the basic way to create an HTTP call. It's very generic/complex there are several examples of how to use it in the codebase.
